x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -652,6 +658,56 @@ public void toXContentFragment(XContentBuilder builder) throws IOException {
         }
     }
 
+    private static final Map<String, CheckedFunction<StreamInput, Object, IOException>> METADATA_VALUE_READER = Map.of(
+        REMOTE_ACCESS_AUTHENTICATION_KEY,
+        Authentication::new,
+        REMOTE_ACCESS_ROLE_DESCRIPTORS_KEY,
+        in -> in.readList(RoleDescriptorsBytes::new)
+    );
```

Comment:
I pondered on this and decided to have this lookup to be the minimal. It means all other map values except this two are still handled with read/write generic value for maximum flexibility. For example, the API key role descriptors can be either bytes or map depending on the version. Or the API key name can be a null. We can handle these known examples. But I feel it's easier to just leave them to be generic values which guarantees to work for known and possibly unknown exceptions.